pkg/controller/base_controller.go

```diff
@@ -0,0 +1,185 @@
+
+	// init function to be executed before reconcile
+	init func() error
+
+	// reconcile is main function, which process the event
+	reconcile func(key string) error
+
+	// destroy is function to be executed for deletion of object
+	destroy func(key string) error
+
+	// controller syncPeriod
+	syncPeriod time.Duration
+
+	cacheSyncWaiters []cache.InformerSynced
+}
+
+// queueOperation represents the type of operation on workQueueLoad
+type queueOperation string
+
+// Different type of operations on the workQueueLoad
+const (
+	qOpDestroy queueOperation = "destroy"
+	qOpSync    queueOperation = "sync"
+)
+
+// workQueueLoad is for storing the key and type of operation before entering workqueue
+type workQueueLoad struct {
+	key       string // Key is the name of the object
+	operation queueOperation
+}
+
+func newController(name string, logger logrus.FieldLogger, numWorker int) *controller {
+	return &controller{
+		name:      name,
+		logger:    logger.WithField("controller", name),
+		workQueue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), name),
+		numWorker: numWorker,
+	}
+}
+
+func (c *controller) Run(ctx context.Context) error {
+	if c.reconcile == nil {
+		return errors.Errorf("no reconcile function provided")
+	}
+
+	if c.init != nil {
+		err := c.init()
+		if err != nil {
+			c.logger.WithError(err).Errorf("init returned error")
+			return err
+		}
+	}
+
+	c.logger.Infof("Starting controller")
+	defer c.logger.Infof("Shutting down controller")
+
+	// wait for caches to sync
+	if len(c.cacheSyncWaiters) > 0 {
+		c.logger.Infof("Waiting for caches to sync")
+		if !cache.WaitForCacheSync(ctx.Done(), c.cacheSyncWaiters...) {
+			return errors.New("timed out waiting for caches to sync")
+		}
+		c.logger.Infof("All caches synced")
+	}
+
+	// Waitgroup for starting controller goroutines.
+	var wg sync.WaitGroup
+
+	wg.Add(c.numWorker)
+	for i := 0; i < c.numWorker; i++ {
+		go func() {
+			wait.Until(c.runWorker, c.syncPeriod, ctx.Done())
+			wg.Done()
+		}()
+	}
+
+	<-ctx.Done()
+	return nil
+}
+
+// runWorker is a long-running function that will continually call the
+// processNextWorkItem function in order to read and process a message on the
+// workqueue.
+func (c *controller) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem will read a single work item off the workqueue and
+// attempt to process it, by calling the syncHandler.
+func (c *controller) processNextWorkItem() bool {
+	key, shutdown := c.workQueue.Get()
+
+	if shutdown {
+		return false
+	}
+
+	// always call Done on this key so the workqueue knows we have finished
+	// processing this item. If any error occurs we re-add this key to workqueue
+	// with rate-limiting.
+	// if we don't want to process this key further then call Forget for this key
+	defer c.workQueue.Done(key)
+
+	qObj, ok := key.(workQueueLoad)
+	if !ok {
+		c.workQueue.Forget(key)
+	}
+
+	var err error
+
+	switch qObj.operation {
+	case qOpDestroy:
```

Comment:
why this separate handling from Sync?
if I'm not wrong, if a CR is deleted when the controller is down, there are chances that qOpDestroy never comes to controller

Reply:
yes. This one is to handle when controller is up. Adding finalizers, which backupcontroller removes, will make sure that this event gets generated.